Christopher Cook Homework 1 Machine Learning BYU

Hello! Today I'm going to open and visualize three interesting datasets. The first one comes from IPUMS usa. They've been collecting ammasing public census data that anyone can use. I've chosen a one percent sample of families from the 1900 US Census.

In [1]:
import pandas as pd

I downloaded the IPUMS data as a .dta so I'll have to use the read_stata function

In [2]:
census = pd.read_stata('C:/Users/Christopher Cook/Downloads/census1900.dta')

In [3]:
census.head(10)

year  datanum  serial    hhwt stateicp                                gq  \
0  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
1  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
2  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
3  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
4  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
5  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
6  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
7  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
8  1900        4    41.0  100.08  Alabama  Households under 1970 definition   
9  1900        4   141.0  100.08  Alabama  Households under 1970 definition   

   pernum   perwt             school                    occ1950  occscore  
0       1  100.08  No, not in school  Truck and tractor drivers        25  
1       2  100.08     Yes, in school                N/A (blank)         0  
2       3  100.08     Yes, in school                N/A (blank)         0  
3       4  100.08  No, not in school         Elevator operators        20  
4       5  100.08  No, not in school          Spinners, textile        18  
5       6  100.08  No, not in school     Charwomen and cleaners        14  
6       7  100.08     Yes, in school                N/A (blank)         0  
7       8  100.08     Yes, in school                N/A (blank)         0  
8       9  100.08  No, not in school                N/A (blank)         0  
9       1  100.08     Yes, in school         Not yet classified         0

This data can help us to see what the US was like in 1900. For instance we can see the percentage of people in school by

In [4]:
result = census['school'].describe()
print (result)
print(result['freq']/result['count'])

count             759981
unique                 2
top       Yes, in school
freq              449677
Name: school, dtype: object
0.591695055534


Note that this percentage is much higher than we would expect for the general population. Since I have a 1% sample of families, I would expect to oversample dependants. Despite this, we can see that education was important for 1900's Americans. This suggests that almost all children were in school.

We can also use this data to explore occupations.

In [5]:
result = census['occ1950'].describe()
print(result)

count          759981
unique            238
top       N/A (blank)
freq           358285
Name: occ1950, dtype: object


We can see that there are 238 unique occupations, the most common of which is blank. This makes sense, as most of our sample is going to school. In fact let's explore that:

In [6]:
census['blank_occ'] = census['occ1950'] == 'N/A (blank)'
census['in_school'] = census['school'] == 'Yes, in school'
print(census.corr())


           datanum        serial          hhwt        pernum         perwt  \
datanum        NaN           NaN           NaN           NaN           NaN   
serial         NaN  1.000000e+00  3.544271e-13  1.186789e-02  3.544271e-13   
hhwt           NaN  3.544271e-13  1.000000e+00 -8.508261e-14  1.000000e+00   
pernum         NaN  1.186789e-02 -8.508261e-14  1.000000e+00 -8.508261e-14   
perwt          NaN  3.544271e-13  1.000000e+00 -8.508261e-14  1.000000e+00   
occscore       NaN -1.346344e-03  4.222617e-14 -1.679467e-01  4.222617e-14   
blank_occ      NaN  1.071978e-02 -9.609727e-14  8.467443e-02 -9.609727e-14   
in_school      NaN  6.023148e-03  4.769329e-14  8.564297e-02  4.769329e-14   

               occscore     blank_occ     in_school  
datanum             NaN           NaN           NaN  
serial    -1.346344e-03  1.071978e-02  6.023148e-03  
hhwt       4.222617e-14 -9.609727e-14  4.769329e-14  
pernum    -1.679467e-01  8.467443e-02  8.564297e-02  
perwt      4.222617e-14 -9.

As we can see, being in school and having a blank occupation are actually negatively correlated. Thus it's more common for students to be listed as students than nothing, or/and that blank occupation people include people from all walks of life--not just students.

Interestingly we can see that having a higher occupation score is associated with not being in school. This come from the obvious fact that providers, with actual, real, jobs, are no longer in school. 

With more variables and time, I could perform some meatier analysis, but this will suffice for now. We move on to my second dataset, a list of all proposed ammendments to the constitution. It comes as a csv

In [7]:
ammend = pd.read_csv('C:/Users/Christopher Cook/Downloads/Proposed_Amendments_from_1787-2014.csv')
ammend.head(6)

identifier source_code  \
0  us-nara-amending-america-000001           A   
1  us-nara-amending-america-000002           A   
2  us-nara-amending-america-000003           A   
3  us-nara-amending-america-000004           A   
4  us-nara-amending-america-000005           A   
5  us-nara-amending-america-000006           A   

                                     source_citation source_index_number  \
0  Ames, Herman V. The Proposed Amendments to the...                   1   
1  Ames, Herman V. The Proposed Amendments to the...                   2   
2  Ames, Herman V. The Proposed Amendments to the...                   3   
3  Ames, Herman V. The Proposed Amendments to the...                   4   
4  Ames, Herman V. The Proposed Amendments to the...                   5   
5  Ames, Herman V. The Proposed Amendments to the...                   6   

                    title_or_description_from_source date_approximation  \
0                 Reservation of nondelegated powers              circa   
1                   Apportionment of Representatives              circa   
2  Restriction on Federal control over election o...              circa   
3       Restriction upon the levying of direct taxes              circa   
4                   Commercial monopolies prohibited              circa   
5                           Indictment by grand jury              circa   

     year  month  day  congress congressional_session  \
0  1788.0    NaN  NaN       NaN                   NaN   
1  1788.0    NaN  NaN       NaN                   NaN   
2  1788.0    NaN  NaN       NaN                   NaN   
3  1788.0    NaN  NaN       NaN                   NaN   
4  1788.0    NaN  NaN       NaN                   NaN   
5  1788.0    NaN  NaN       NaN                   NaN   

  joint_resolution_chamber  joint_resolution_number sponsor_name  \
0                      NaN                      NaN          NaN   
1                      NaN                      NaN          NaN   
2                      NaN                      NaN          NaN   
3                      NaN                      NaN          NaN   
4                      NaN                      NaN          NaN   
5                      NaN                      NaN          NaN   

  sponsor_state_or_territory committee_of_referral  
0                        NaN                   NaN  
1                        NaN                   NaN  
2                        NaN                   NaN  
3                        NaN                   NaN  
4                        NaN                   NaN  
5                        NaN                   NaN

This allows us to see the frequency of US ammendments across time.

In [8]:
import matplotlib.pyplot as plt
import numpy as np

In [9]:
print(ammend['year'].describe())



count    11792.000000
mean      1942.255597
std         51.634764
min       1788.000000
25%       1919.000000
50%       1963.000000
75%       1975.000000
max       2014.000000
Name: year, dtype: float64


We can also find out the average number of ammendements per year, and the standard deviation per year

In [10]:
years = {}
for x in range(1776, 2017):
    years[x] = 0

for index, row in ammend.iterrows():
    if row['year'] in years:
        years[row['year']] += 1
    else:
        years[row['year']] = 1  
years = pd.DataFrame(years, index=range(0,1))


In [11]:
years.transpose().describe()

0
count  246.000000
mean    47.955285
std     79.891521
min      0.000000
25%      7.000000
50%     20.000000
75%     51.750000
max    616.000000

So, on average there have been 48 ammendments per year, with a standard deviation of 80! Apparently there was one year where there were 616 ammendments proposed! That's a lot of ammendments.

I could do some text analysis on the kinds of ammendments getting proposed, but I'll leave that to my final dataset, a bunch of jokes!

In [12]:
import json
with open('C:/Users/Christopher Cook/Downloads/jay.txt') as f:
    jokes = json.load(f)
    

In [13]:
jokes['value'][0].keys()


dict_keys(['id', 'joke', 'categories'])

Since I've managed to load this json, let's see if we can get it into a dataframe


In [14]:
dfj = pd.DataFrame(jokes['value'])

In [15]:
dfj.head(5)

categories   id                                               joke
0         []  335  Chuck Norris likes his coffee half and half: h...
1         []   21  Chuck Norris doesn't shower, he only takes blo...
2         []  531  Jesus can walk on water, but Chuck Norris can ...
3         []  347  Some people ask for a Kleenex when they sneeze...
4         []  358  The 11th commandment is ?Thou shalt not piss o...

We can now conduct frivolous analysis on this dataset. For example, what is joke number 21?

In [16]:
dfj.at[21,'joke']

"Chuck Norris' house has no doors, only walls that he walks through."

Let's see... How many characters, is the average joke?

In [17]:
dfj['length'] = dfj['joke'].str.len()
dfj['length'].describe()

count     31.000000
mean      91.419355
std       37.896151
min       40.000000
25%       63.500000
50%       83.000000
75%      113.000000
max      211.000000
Name: length, dtype: float64

What about number of words?

In [18]:
dfj['words'] = dfj['joke'].str.split()

In [19]:
dfj['num_words'] = dfj['words'].str.len()
dfj['num_words'].describe()

count    31.000000
mean     16.225806
std       7.050814
min       6.000000
25%      12.000000
50%      15.000000
75%      20.000000
max      39.000000
Name: num_words, dtype: float64

We can even look at average word length

In [20]:
dfj['avg_len'] = dfj['length'] / dfj['num_words']
dfj['avg_len'].describe()

count    31.000000
mean      5.753256
std       0.783032
min       4.705882
25%       5.333333
50%       5.600000
75%       6.021739
max       8.714286
Name: avg_len, dtype: float64

Though this kind of analysis won't tell us how funny a joke is, we can go through and mark them as either funny or not, which could potentially be used for as a machine learning training set